compiler/src/iree/compiler/Codegen/Common/TileAndFuseUtils.cpp

```diff
@@ -106,30 +111,54 @@ void collectTiledAndFusedOps(Operation *rootOp,
 }
 
 FailureOr<std::queue<Operation *>>
-fuseConsumersIntoForall(RewriterBase &rewriter, Operation *tiledOp,
-                        MutableArrayRef<LoopLikeOpInterface> loops,
-                        bool useWARForConsumerFusionSSAViolation) {
-  auto addCandidateSlices =
-      [](Operation *fusedOp,
-         std::queue<tensor::ParallelInsertSliceOp> &candidates) {
-        for (auto *userOp : fusedOp->getResults().getUsers()) {
-          if (auto sliceOp =
-                  llvm::dyn_cast<tensor::ParallelInsertSliceOp>(userOp)) {
-            candidates.push(sliceOp);
-          }
-        }
-      };
+fuseConsumersIntoLoops(RewriterBase &rewriter, Operation *tiledOp,
+                       MutableArrayRef<LoopLikeOpInterface> loops,
+                       bool useWARForConsumerFusionSSAViolation) {
+  auto addCandidateSlices = [](Operation *fusedOp,
+                               std::queue<Operation *> &candidates) {
+    for (auto *userOp : fusedOp->getResults().getUsers()) {
+      if (llvm::isa<tensor::InsertSliceOp, tensor::ParallelInsertSliceOp>(
+              userOp)) {
+        // Users of tiledOp should either be all of type `tensor.insert_slice`
+        // or all of`tensor.parallel_insert_slice`.
+        //
+        // Pattern 1 - tileing with scf.for:
+        //   %out = scf.for ... {
+        //     %0 = scf.for ... {
+        //       %t0 = op
+        //       %t1 = op  %t0                 // <- `tiledOp`
+        //       %1 = tensor.insert_slice %t1
+        //       yield %1
+        //     }
+        //     yield %0
+        //   }
+        //
+        // Pattern 2 - tiling with scf.forall:
+        //   % out = scf.forall ... {
+        //       %t0 = op
+        //       %t1 = op  %t0                 // <- `tiledOp`
+        //       scf.forall.in_parallel {
+        //         tensor.parallel_insert_slice %tile
+        //       }
+        //   }
+        assert(candidates.empty() ||
+               candidates.front()->getName() == userOp->getName() &&
+                   "expected all slice users to be of type tensor.insert_slice "
+                   "or of tensor.parallel_insert_slice.");
```

Comment:
The order matters for some cases. We had hit similar bug reports before, so I'd suggest to always wrap the conditions within braces, if there are more than one conditions.

```suggestion
        assert((candidates.empty() ||
               candidates.front()->getName() == userOp->getName()) &&
                   "expected all slice users to be of type tensor.insert_slice "
                   "or of tensor.parallel_insert_slice.");
```